app/src/org/commcare/android/util/MediaUtil.java

```diff
@@ -1,82 +1,274 @@
-            String latitude = Double.toString(mGeoPointData.getValue()[0]);
-            String longitude= Double.toString(mGeoPointData.getValue()[1]);
-            return "geo:" + latitude + "," + longitude + "?q=" + latitude + "," + longitude;
-            
-        }catch(IllegalArgumentException iae){
-            return "geo:0,0?q=" + rawInput;
+    public static Bitmap getBitmapScaledForNativeDensity(DisplayMetrics metrics, String imageFilepath,
+                                                         int containerHeight, int containerWidth,
+                                                         int targetDensity) {
+        BitmapFactory.Options o = new BitmapFactory.Options();
+        o.inJustDecodeBounds = true;
+        o.inScaled = false;
+        BitmapFactory.decodeFile(imageFilepath, o);
+        int imageHeight = o.outHeight;
+        int imageWidth = o.outWidth;
+
+        double scaleFactor = computeInflationScaleFactor(metrics, targetDensity);
+        int calculatedHeight = Math.round((float)(imageHeight * scaleFactor));
+        int calculatedWidth = Math.round((float)(imageWidth * scaleFactor));
+
+        if (containerHeight < imageHeight || containerWidth < imageWidth || calculatedHeight < imageHeight) {
+            // If either the container dimens or calculated dimens impose a smaller dimension,
+            // scale down
+            return getBitmapScaledDownExact(imageFilepath, imageHeight, imageWidth,
+                    calculatedHeight, calculatedWidth, containerHeight, containerWidth);
+        } else {
+            return attemptBoundedScaleUp(imageFilepath, calculatedHeight, calculatedWidth,
+                    containerHeight, containerWidth);
         }
     }
-	
-	public static String stripArguments(String input){
-		if(input.contains("{") && input.contains("}")){
-			String replaced = input.substring(input.indexOf("{")-1, input.indexOf("}")+1);
-			return input.replace(replaced, "").trim();
-		}
-		return input;
-	}
+
+    /**
+     * @return A bitmap representation of the given image file, scaled down such that the new
+     * dimensions of the image is the SMALLER of the following 2 options:
+     * 1) newCalcHeight and newCalcWidth
+     * 2) the largest dimensions for which the original aspect ratio is maintained, without
+     * exceeding either boundingWidth or boundingHeight
+     */
+    private static Bitmap getBitmapScaledDownExact(String imageFilepath,
+                                                   int originalHeight, int originalWidth,
+                                                   int calcHeight, int calcWidth,
+                                                   int boundingHeight, int boundingWidth) {
+
+        Pair<Integer, Integer> dimensImposedByContainer = getProportionalDimensForContainer(
+                originalHeight, originalWidth, boundingHeight, boundingWidth);
+        int newWidth = Math.min(dimensImposedByContainer.first, calcWidth);

```

Comment:
Since the values of both calcWidth/calcHeight, and dimensImposedByContainer.first/second preserve the aspect ratio, it's not possible that you would end up with dimens from different pairs getting mixed.